### PR TITLE
Feature/api 9 list of available parcels

### PR DESCRIPTION
**Pull request by Asad-noor**
API [BE] list of available parcels

### CODE REVIEW

File: internal/app/carrier/service_test.go [hunk 1]
```diff
@@ -4,7 +4,7 @@ import (
 	"context"
 	"errors"
 	"parcel-service/internal/app/model"
-	"parcel-service/internal/app/service/mocks"
+	mock_service "parcel-service/internal/app/service/mocks"
```
Comment (rdhawladar) [inline internal/app/carrier/service_test.go]: remove mock_service

File: internal/app/parcel/repository.go [hunk 2]
```diff
@@ -48,6 +49,19 @@ func (r *repository) InsertParcel(ctx context.Context, parcel model.Parcel) erro
 	return nil
 }
 
+func (r *repository) GetParcelsList(ctx context.Context, status int, limit int, offset int) ([]model.Parcel, error) {
+	var parcels []model.Parcel
+	//r.db.Select(&parcels, getParcelListQuery)
```
Comment (rdhawladar) [inline internal/app/parcel/repository.go]: Remove commented code.

File: internal/app/carrier/service_test.go [hunk 3]
```diff
@@ -23,14 +23,14 @@ func TestService_NewCarrierRequest(t *testing.T) {
 	testCases := []struct {
 		desc     string
 		payload  model.CarrierRequest
-		mockRepo func() *mocks.MockCarrierRequestRepository
+		mockRepo func() *mock_service.MockCarrierRepository
```
Comment (rdhawladar) [inline internal/app/carrier/service_test.go]: change it to mocks

File: internal/app/parcel/repository_test.go [hunk 4]
```diff
@@ -67,6 +67,73 @@ func TestRepository_InsertParcel(t *testing.T) {
...
+	t.Run("should return error", func(t *testing.T) {
+		db, m, _ := sqlmock.New()
+		defer db.Close()
+
+		sqlxDB := sqlx.NewDb(db, "sqlmock")
+		m.ExpectQuery("^SELECT (.+) FROM parcel WHERE (.+)").
+			WithArgs(3).
+			WillReturnError(errors.New("sql-error"))
+		repo := NewRepository(sqlxDB)
+		_, err := repo.GetParcelsList(context.Background(), 0, 0, 0)
```
Comment (rdhawladar) [inline internal/app/parcel/repository_test.go]: Please add test case for empty data when it returns 0 row.

File: internal/app/parcel/service.go [hunk 5]
```diff
@@ -12,12 +12,23 @@ type service struct {
 	repo svc.ParcelRepository
 }
 
+// NewService is to generate for new repo
```
Comment (rdhawladar) [inline internal/app/parcel/service.go]: Remove single line comments.

File: internal/app/parcel/service.go [hunk 6]
```diff
@@ -12,12 +12,23 @@ type service struct {
...
 	return &service{
 		repo: repo,
 	}
 }
 
+// get Parcel List
+func (s *service) GetParcels(ctx context.Context, status int, limit int, offset int) ([]model.Parcel, error) {
+	parcel, err := s.repo.GetParcelsList(ctx, status, limit, offset)
+
+	if err != nil {
```
Comment (rdhawladar) [inline internal/app/parcel/service.go]: No need to check error here. This error is coming from repository, just return the output of the repo to handler.

File: internal/app/server/handler.go [hunk 7]
```diff
@@ -11,6 +12,34 @@ import (
...
+	limit, err := strconv.Atoi(r.URL.Query().Get("limit"))
+	if err == nil {
+		fmt.Printf("%d", limit)
+	}
+
+	parcel, err := s.parcelService.GetParcels(r.Context(), status, limit, offset)
+	if err != nil {
+		// if errors.Is(err, model.ErrInvalid) {
+		// 	ErrInvalidEntityResponse(w, "invalid parcel", err)
+		// 	return
```
Comment (rdhawladar) [inline internal/app/server/handler.go]: Remove cemented code.

File: internal/app/server/handler.go [hunk 8]
```diff
@@ -11,6 +12,34 @@ import (
...
+		// }
+		fmt.Print(err)
+		log.Error().Err(err).Msgf("[parcel/{id}] failed to parcel '%d': %v", nil, err)
```
Comment (rdhawladar) [inline internal/app/server/handler.go]: add the function name instead of route.

File: internal/app/server/handler_test.go [hunk 9]
```diff
@@ -248,11 +333,10 @@ func TestGetParcel(t *testing.T) {
 
 			w := httptest.NewRecorder()
 			body := strings.NewReader("")
-			r := httptest.NewRequest(http.MethodGet, "/api/v1/parcel/"+tc.parcelID, body)
-			r = mux.SetURLVars(r, map[string]string{"id": tc.parcelID})
+			r := httptest.NewRequest(http.MethodGet, "/api/v1/parcel/"+strconv.Itoa(tc.status)+strconv.Itoa(tc.offset)+strconv.Itoa(tc.limit), body)
```
Comment (rdhawladar) [inline internal/app/server/handler_test.go]: I think "strconv.Itoa(tc.status)+strconv.Itoa(tc.offset)+strconv.Itoa(tc.limit" this part has no meaning here. It passes the wrong url like '/api/v1/parcel/1/20/20' . This is not our endpoint.

File: internal/app/parcel/repository.go [hunk 10]
```diff
@@ -14,6 +14,7 @@ import (
 // SQL Query and error
 const (
 	errUniqueViolation   = pq.ErrorCode("23505")
+	getParcelListQuery   = `SELECT * FROM parcel WHERE status=$1 LIMIT $2 OFFSET $3`
```
Comment (shuza) [inline internal/app/parcel/repository.go]: don't use `*` mention exact fields that you want to fetch

File: internal/app/parcel/repository.go [hunk 11]
```diff
@@ -48,6 +49,19 @@ func (r *repository) InsertParcel(ctx context.Context, parcel model.Parcel) erro
 	return nil
 }
 
+func (r *repository) GetParcelsList(ctx context.Context, status int, limit int, offset int) ([]model.Parcel, error) {
+	var parcels []model.Parcel
+	//r.db.Select(&parcels, getParcelListQuery)
+	if err := r.db.Select(&parcels, getParcelListQuery, status, limit, offset); err != nil {
```
Comment (shuza) [inline internal/app/parcel/repository.go]: you need to use method which is bind with context otherwise query will not cancel when context close or remove

File: internal/app/parcel/repository.go [hunk 12]
```diff
@@ -48,6 +49,19 @@ func (r *repository) InsertParcel(ctx context.Context, parcel model.Parcel) erro
...
+		if pqErr, ok := err.(*pq.Error); ok && pqErr.Code == errUniqueViolation {
```
Comment (shuza) [inline internal/app/parcel/repository.go]: why need to check unique key violation at here???

File: internal/app/parcel/service.go (hunk 6 above)
Comment (shuza) [reply internal/app/parcel/service.go]: agree with @rdhawladar if you don't do any business depending on the error then you can directly return it

File: internal/app/parcel/service_test.go [hunk 13]
```diff
@@ -16,6 +17,111 @@ const (
...
+			expErr:    nil,
+			expParcel: []model.Parcel{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			s := NewService(tc.mockRepo())
+			parcels, err := s.GetParcels(context.Background(), 1, 2, 4)
+			fmt.Println(parcels)
```
Comment (shuza) [inline internal/app/parcel/service_test.go]: remove those prints

File: internal/app/parcel/service_test.go [hunk 14]
```diff
@@ -16,6 +17,111 @@ const (
...
+			fmt.Println("heloooo")
```
Comment (shuza) [inline internal/app/parcel/service_test.go]: please don't push your test code

File: internal/app/parcel/service_test.go [hunk 15]
```diff
@@ -16,6 +17,111 @@ const (
...
+			fmt.Println(tc.expParcel)
+			fmt.Println("heloooo    2")
```
Comment (shuza) [inline internal/app/parcel/service_test.go]: please don't push your test code

File: internal/app/parcel/service_test.go [hunk 16]
```diff
@@ -16,6 +17,111 @@ const (
...
+			assert.Equal(t, tc.expErr, err)
+			assert.EqualValues(t, tc.expParcel, parcels)
```
Comment (shuza) [inline internal/app/parcel/service_test.go:127]: if there is error then you can check nil for parcels but don't use EqualValues until you really need to check the values of the object

Comment (shuza) [reply internal/app/parcel/service_test.go:127]: its a waste of resource

File: internal/app/server/handler.go [hunk 17]
```diff
@@ -11,6 +12,34 @@ import (
...
 )
 
+func (s *server) getParcelList(w http.ResponseWriter, r *http.Request) {
+	status, err := strconv.Atoi(r.URL.Query().Get("status"))
+	if err == nil {
+		fmt.Printf("%d", status)
+	}
+	offset, err := strconv.Atoi(r.URL.Query().Get("offset"))
+	if err == nil {
+		fmt.Printf("%d", offset)
```
Comment (shuza) [inline internal/app/server/handler.go]: you should return invalid error if offset is not valid

File: internal/app/server/handler.go [hunk 18]
```diff
@@ -11,6 +12,34 @@ import (
...
+	}
+	limit, err := strconv.Atoi(r.URL.Query().Get("limit"))
+	if err == nil {
+		fmt.Printf("%d", limit)
```
Comment (shuza) [inline internal/app/server/handler.go]: you should return invalid error if limit is not valid

File: internal/app/server/handler_test.go [hunk 19]
```diff
@@ -88,6 +89,105 @@ func TestNewParcel(t *testing.T) {
...
+		{
+			desc: "should return internal server error",
+			mockParcelSvc: func() *mock_service.MockParcelService {
+				s := mock_service.NewMockParcelService(ctrl)
+				s.EXPECT().GetParcelByID(gomock.Any(), 1).
+					Return(model.Parcel{}, model.ErrNotFound)
+				return s
+			},
+			parcelID:      "1",
+			expStatusCode: http.StatusBadRequest,
```
Comment (shuza) [inline internal/app/server/handler_test.go]: it should be not found error not bad request

File: internal/app/server/handler_test.go [hunk 20]
```diff
@@ -88,6 +89,105 @@ func TestNewParcel(t *testing.T) {
...
+				s.EXPECT().GetParcelByID(gomock.Any(), 1).
+					Return(model.Parcel{}, errors.New("server-error"))
+				return s
+			},
+			parcelID:      "1",
+			expStatusCode: http.StatusInternalServerError,
+			expResponse:   `{"success":false,"errors":[{"code":"SERVER_ERROR","message":"server-error","message_title":"Failed to fetch parcel 1","severity":"error"}],"data":null}`,
+		},
+		{
+			desc: "should return internal server error",
```
Comment (shuza) [inline internal/app/server/handler_test.go]: change the description according to test case

File: internal/app/server/handler_test.go (hunk 9 above)
Comment (shuza) [reply internal/app/server/handler_test.go]: better to build the url string

**Review by shuza — CHANGES_REQUESTED** [review]
Please do push your debugging code. its a waste of time for reviewer. Also address those comments. it seems like an amature PR

File: internal/app/parcel/repository_test.go (hunk 4 above)
Comment (Asad-noor) [reply internal/app/parcel/repository_test.go]: This test case will return 0 row. Let me know if you have any confusion.

File: internal/app/parcel/repository.go (hunk 10 above)
Comment (Asad-noor) [reply internal/app/parcel/repository.go]: If I want all the fields in that case?

File: internal/app/parcel/repository.go (hunk 11 above)
Comment (Asad-noor) [reply internal/app/parcel/repository.go]: I didn't understand, please provide a proper example.

Comment (Asad-noor) [reply internal/app/parcel/repository.go]: if err := r.db.GetContext(ctx, &parcels, getParcelListQuery, status, limit, offset);

is it what you mean by bind to context?

File: internal/app/server/handler_test.go (hunk 19 above)
Comment (Asad-noor) [reply internal/app/server/handler_test.go]: not my code though I have fixed on mine.

File: internal/app/server/handler_test.go (hunk 20 above)
Comment (Asad-noor) [reply internal/app/server/handler_test.go]: not my code though I have fixed on mine.

File: internal/app/parcel/repository_test.go (hunk 4 above)
Comment (rdhawladar) [reply internal/app/parcel/repository_test.go]: I dont have confusion but you have. I asked to add a test case for empty daata, such as you request for parcel_id 100 and it doesn't exist on DB. You need to check this case.

File: internal/app/server/handler_test.go (hunk 20 above)
Comment (shuza) [reply internal/app/server/handler_test.go]: it comes with your PR so definitely its your code

File: internal/app/server/handler_test.go (hunk 19 above)
Comment (shuza) [reply internal/app/server/handler_test.go]: it comes with your PR so definitely its your code

**Review by shuza — CHANGES_REQUESTED** [review]
Any code comes with your PR and you should be responsible for that

File: internal/app/server/handler_test.go (hunk 19 above)
Comment (Asad-noor) [reply internal/app/server/handler_test.go]: It comes after merging with master so don't be confused with that.

File: internal/app/carrier/service_test.go (hunk 1 above)
Comment (Asad-noor) [reply internal/app/carrier/service_test.go]: done

File: internal/app/parcel/repository.go (hunk 2 above)
Comment (Asad-noor) [reply internal/app/parcel/repository.go]: done

File: internal/app/carrier/service_test.go (hunk 3 above)
Comment (Asad-noor) [reply internal/app/carrier/service_test.go]: done

Comment (Asad-noor) [reply internal/app/carrier/service_test.go]: done

File: internal/app/parcel/repository_test.go (hunk 4 above)
Comment (Asad-noor) [reply internal/app/parcel/repository_test.go]: done

File: internal/app/parcel/service.go (hunk 5 above)
Comment (Asad-noor) [reply internal/app/parcel/service.go]: done

File: internal/app/parcel/service.go (hunk 6 above)
Comment (Asad-noor) [reply internal/app/parcel/service.go]: done

File: internal/app/server/handler.go (hunk 7 above)
Comment (Asad-noor) [reply internal/app/server/handler.go]: done

File: internal/app/server/handler.go (hunk 8 above)
Comment (Asad-noor) [reply internal/app/server/handler.go]: done

File: internal/app/server/handler_test.go (hunk 9 above)
Comment (Asad-noor) [reply internal/app/server/handler_test.go]: done

File: internal/app/parcel/repository.go (hunk 12 above)
Comment (Asad-noor) [reply internal/app/parcel/repository.go]: removed

File: internal/app/parcel/service_test.go (hunk 13 above)
Comment (Asad-noor) [reply internal/app/parcel/service_test.go]: done

File: internal/app/server/handler.go (hunk 17 above)
Comment (Asad-noor) [reply internal/app/server/handler.go]: done

File: internal/app/server/handler.go (hunk 18 above)
Comment (Asad-noor) [reply internal/app/server/handler.go]: done

File: internal/app/server/handler_test.go (hunk 19 above)
Comment (shuza) [reply internal/app/server/handler_test.go]: i am not confused. any code changed by your PR its your code. if it was in master then. it should not show change

**Review by shuza — CHANGES_REQUESTED** [review]
Please resolve all comments before you request for another review

File: internal/app/parcel/repository.go (hunk 10 above)
Comment (zrshishir) [reply internal/app/parcel/repository.go]: generally there are some fields that we don't need as created_at, updated_at, status data and all admin related data. If we don't need the why should we pull all data? It is a good practice to pull only required data. There are lot of benefits. 

-  Front end implementation extra data harassment relief 
-  api documentation and QA if there is extra data field you will have to give a good reason to keep the data.
- Finally, when you work on a open source project then it totally illogical to get irrelevant data. 

There are lot more things to be counted.

File: internal/app/parcel/repository_test.go [hunk 21]
```diff
@@ -67,7 +68,83 @@ func TestRepository_InsertParcel(t *testing.T) {
...
+
+		sqlxDB := sqlx.NewDb(db, "sqlmock")
+
+		m.ExpectQuery(regexp.QuoteMeta("SELECT * FROM parcel WHERE status=$1 LIMIT $2 OFFSET $3")).
+			WillReturnRows(sqlmock.NewRows([]string{"user_id", "status", "source_address", "destination_address", "type", "price", "carrier_fee", "company_fee"}).
+				AddRow(parcels[0].UserID, parcels[0].Status, parcels[0].SourceAddress, parcels[0].DestinationAddress, parcels[0].ParcelType, parcels[0].Price, parcels[0].CarrierFee, parcels[0].CompanyFee).
+				AddRow(parcels[1].UserID, parcels[1].Status, parcels[1].SourceAddress, parcels[1].DestinationAddress, parcels[1].ParcelType, parcels[1].Price, parcels[1].CarrierFee, parcels[1].CompanyFee))
+
+		repo := NewRepository(sqlxDB)
+		result, err := repo.GetParcelsList(context.Background(), 1, 2, 0)
```
Comment (zrshishir) [inline internal/app/parcel/repository_test.go]: It will be good for me to understand var name instead of value. In line 113 => 1, 2, 0 replace it with var name.

File: internal/app/parcel/repository_test.go [hunk 22]
```diff
@@ -67,7 +68,83 @@ func TestRepository_InsertParcel(t *testing.T) {
...
+		db, m, _ := sqlmock.New()
+		defer db.Close()
+
+		sqlxDB := sqlx.NewDb(db, "sqlmock")
+		m.ExpectQuery("^SELECT (.+) FROM parcel WHERE (.+)").
+			WithArgs(0, 0, 0).
+			WillReturnRows(sqlmock.NewRows([]string{}))
+
+		repo := NewRepository(sqlxDB)
+		res, err := repo.GetParcelsList(context.Background(), 0, 0, 0)
```
Comment (zrshishir) [inline internal/app/parcel/repository_test.go]: Same here, It will be good for me to understand var name instead of value.

File: internal/app/parcel/repository_test.go [hunk 23]
```diff
@@ -67,7 +68,83 @@ func TestRepository_InsertParcel(t *testing.T) {
...
+		db, m, _ := sqlmock.New()
+		defer db.Close()
+
+		sqlxDB := sqlx.NewDb(db, "sqlmock")
+		m.ExpectQuery("^SELECT (.+) FROM parcel WHERE (.+)").
+			WithArgs(0, 0, -1).
+			WillReturnError(sql.ErrNoRows)
+
+		repo := NewRepository(sqlxDB)
+		_, err := repo.GetParcelsList(context.Background(), 0, 0, -1)
```
Comment (zrshishir) [inline internal/app/parcel/repository_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/parcel/service_test.go [hunk 24]
```diff
@@ -24,6 +24,107 @@ var parcel = model.Parcel{
...
+		expParcel []model.Parcel
+	}{
+		{
+			desc:   "should return success",
+			status: 1,
+			offset: 2,
+			limit:  4,
+			mockRepo: func() *mocks.MockParcelRepository {
+				r := mocks.NewMockParcelRepository(ctrl)
+				r.EXPECT().GetParcelsList(gomock.Any(), 1, 2, 4).Return(parcels, nil)
```
Comment (zrshishir) [inline internal/app/parcel/service_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/parcel/service_test.go [hunk 25]
```diff
@@ -24,6 +24,107 @@ var parcel = model.Parcel{
...
+		},
+
+		{
+			desc:   "Should return not found",
+			status: 1,
+			offset: 2,
+			limit:  4,
+			mockRepo: func() *mocks.MockParcelRepository {
+				r := mocks.NewMockParcelRepository(ctrl)
+				r.EXPECT().GetParcelsList(gomock.Any(), 1, 2, 4).Return([]model.Parcel{}, model.ErrNotFound)
```
Comment (zrshishir) [inline internal/app/parcel/service_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/parcel/service_test.go [hunk 26]
```diff
@@ -24,6 +24,107 @@ var parcel = model.Parcel{
...
+			expParcel: []model.Parcel(nil),
+		},
+		{
+			desc:   "should return DB error",
+			status: 1,
+			offset: 2,
+			limit:  4,
+			mockRepo: func() *mocks.MockParcelRepository {
+				r := mocks.NewMockParcelRepository(ctrl)
+				r.EXPECT().GetParcelsList(gomock.Any(), 1, 2, 4).Return([]model.Parcel{}, errors.New("db-error"))
```
Comment (zrshishir) [inline internal/app/parcel/service_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/parcel/service_test.go [hunk 27]
```diff
@@ -24,6 +24,107 @@ var parcel = model.Parcel{
...
+			expParcel: []model.Parcel(nil),
+		},
+		{
+			desc:   "should return empty parcel",
+			status: 1,
+			offset: 2,
+			limit:  4,
+			mockRepo: func() *mocks.MockParcelRepository {
+				r := mocks.NewMockParcelRepository(ctrl)
+				r.EXPECT().GetParcelsList(gomock.Any(), 1, 2, 4).Return([]model.Parcel{}, nil)
```
Comment (zrshishir) [inline internal/app/parcel/service_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/parcel/service_test.go [hunk 28]
```diff
@@ -24,6 +24,107 @@ var parcel = model.Parcel{
...
+			},
+			expErr:    nil,
+			expParcel: []model.Parcel{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			s := NewService(tc.mockRepo())
+			parcels, err := s.GetParcels(context.Background(), 1, 2, 4)
```
Comment (zrshishir) [inline internal/app/parcel/service_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/server/handler_test.go [hunk 29]
```diff
@@ -88,6 +88,105 @@ func TestNewParcel(t *testing.T) {
...
+		mockParcelSvc func() *mocks.MockParcelService
+		parcelID      string
+		expStatusCode int
+		expResponse   string
+	}{
+		{
+			desc: "should success",
+			mockParcelSvc: func() *mocks.MockParcelService {
+				s := mocks.NewMockParcelService(ctrl)
+				s.EXPECT().GetParcelByID(gomock.Any(), 1).Return(parcel, nil)
```
Comment (zrshishir) [inline internal/app/server/handler_test.go]: use parcelID instead of 1. It will be good for me to understand var name instead of value.

File: internal/app/server/handler_test.go [hunk 30]
```diff
@@ -88,6 +88,105 @@ func TestNewParcel(t *testing.T) {
...
+			},
+			parcelID:      "1",
+			expStatusCode: http.StatusOK,
+			expResponse:   `{"success":true,"errors":null,"data":{"id":0,"user_id":1,"carrier_id":0,"status":0,"source_address":"Dhaka Bangladesh","destination_address":"Pabna Shadar","source_time":"2020-04-11T21:34:01Z","type":"Document","price":200,"carrier_fee":180,"company_fee":20,"created_at":"2020-04-11T21:34:01Z","updated_at":"2020-04-11T21:34:01Z"}}`,
+		},
+		{
+			desc: "should return ID not exist",
+			mockParcelSvc: func() *mocks.MockParcelService {
+				s := mocks.NewMockParcelService(ctrl)
+				s.EXPECT().GetParcelByID(gomock.Any(), 1).Return(model.Parcel{}, model.ErrInvalid)
```
Comment (zrshishir) [inline internal/app/server/handler_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/server/handler_test.go [hunk 31]
```diff
@@ -88,6 +88,105 @@ func TestNewParcel(t *testing.T) {
...
+			},
+			parcelID:      "1",
+			expStatusCode: http.StatusBadRequest,
+			expResponse:   `{"success":false,"errors":[{"code":"INVALID","message":"invalid","message_title":"This ID does not exist.","severity":"error"}],"data":null}`,
+		},
+		{
+			desc: "should return internal server error",
+			mockParcelSvc: func() *mocks.MockParcelService {
+				s := mocks.NewMockParcelService(ctrl)
+				s.EXPECT().GetParcelByID(gomock.Any(), 1).
```
Comment (zrshishir) [inline internal/app/server/handler_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/server/handler_test.go (hunk 20 above)
Comment (zrshishir) [reply internal/app/server/handler_test.go]: When you use a code or logic or code segment on your branch. You will have to take the responsibility for the code and you should understand clearly before use it. You can't avoid your responsibility to say "not my code".

File: internal/app/server/handler_test.go [hunk 32]
```diff
@@ -88,6 +88,105 @@ func TestNewParcel(t *testing.T) {
...
+			},
+			parcelID:      "1",
+			expStatusCode: http.StatusInternalServerError,
+			expResponse:   `{"success":false,"errors":[{"code":"SERVER_ERROR","message":"server-error","message_title":"Failed to fetch parcel 1","severity":"error"}],"data":null}`,
+		},
+		{
+			desc: "should return internal server error",
+			mockParcelSvc: func() *mocks.MockParcelService {
+				s := mocks.NewMockParcelService(ctrl)
+				s.EXPECT().GetParcelByID(gomock.Any(), 1).
```
Comment (zrshishir) [inline internal/app/server/handler_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/server/handler_test.go (hunk 19 above)
Comment (zrshishir) [reply internal/app/server/handler_test.go]: If there is something wrong on master or other branch that we didn't notice before, we should correct them whenever it is seen.

File: internal/app/server/handler_test.go [hunk 33]
```diff
@@ -161,84 +260,86 @@ func TestAddCarrierRequest(t *testing.T) {
...
+		limit         string
 		expStatusCode int
 		expResponse   string
 	}{
 		{
 			desc: "should success",
 			mockParcelSvc: func() *mocks.MockParcelService {
 				s := mocks.NewMockParcelService(ctrl)
-				s.EXPECT().GetParcelByID(gomock.Any(), 1).Return(parcel, nil)
+				s.EXPECT().GetParcels(gomock.Any(), 1, 2, 0).Return(parcels, nil)
```
Comment (zrshishir) [inline internal/app/server/handler_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/server/handler_test.go [hunk 34]
```diff
@@ -161,84 +260,86 @@ func TestAddCarrierRequest(t *testing.T) {
...
+			expResponse:   `{"success":true,"errors":null,"data":[{"id":0,"user_id":1,"carrier_id":0,"status":1,"source_address":"Dhaka Bangladesh","destination_address":"Pabna Shadar","source_time":"2020-04-11T21:34:01Z","type":"Document","price":200,"carrier_fee":180,"company_fee":20,"created_at":"2020-04-11T21:34:01Z","updated_at":"2020-04-11T21:34:01Z"},{"id":0,"user_id":2,"carrier_id":0,"status":1,"source_address":"Dhaka Bangladesh","destination_address":"Pabna Shadar","source_time":"2020-04-11T21:34:01Z","type":"Document","price":200,"carrier_fee":180,"company_fee":20,"created_at":"2020-04-11T21:34:01Z","updated_at":"2020-04-11T21:34:01Z"}]}`,
 		},
 		{
-			desc: "should return ID not exist",
+			desc: "should return empty parcel list",
 			mockParcelSvc: func() *mocks.MockParcelService {
 				s := mocks.NewMockParcelService(ctrl)
-				s.EXPECT().GetParcelByID(gomock.Any(), 1).Return(model.Parcel{}, model.ErrInvalid)
+				s.EXPECT().GetParcels(gomock.Any(), 0, 2, 0).Return(nil, nil)
```
Comment (zrshishir) [inline internal/app/server/handler_test.go]: It will be good for me to understand var name instead of value.

File: internal/app/server/handler_test.go [hunk 35]
```diff
@@ -161,84 +260,86 @@ func TestAddCarrierRequest(t *testing.T) {
...
+			expStatusCode: http.StatusOK,
+			expResponse:   `{"success":true,"errors":null,"data":null}`,
 		},
 		{
 			desc: "should return internal server error",
 			mockParcelSvc: func() *mocks.MockParcelService {
 				s := mocks.NewMockParcelService(ctrl)
-				s.EXPECT().GetParcelByID(gomock.Any(), 1).
-					Return(model.Parcel{}, errors.New("server-error"))
+				s.EXPECT().GetParcels(gomock.Any(), 1, 2, -1).Return([]model.Parcel{}, errors.New("pq: OFFSET must not be negative"))
```
Comment (zrshishir) [inline internal/app/server/handler_test.go]: It will be good for me to understand var name instead of value.

**Review by zrshishir — CHANGES_REQUESTED** [review]
Some basic things to change, please look over it.

File: internal/app/parcel/repository_test.go (hunk 21 above)
Comment (Asad-noor) [reply internal/app/parcel/repository_test.go]: Please provide a proper example for all the test cases, as I can't use same variable initialization for all cases.

**Review by rdhawladar — APPROVED** [review]
Looks good to me.

File: internal/app/parcel/repository.go [hunk 36]
```diff
@@ -49,6 +50,18 @@ func (r *repository) InsertParcel(ctx context.Context, parcel model.Parcel) erro
...
 }
 
+func (r *repository) GetParcelsList(ctx context.Context, status int, limit int, offset int) ([]model.Parcel, error) {
+	var parcels []model.Parcel
+	if err := r.db.SelectContext(ctx, &parcels, getParcelListQuery, status, limit, offset); err != nil {
+		if err == sql.ErrNoRows {
+			log.Error().Err(err).Msgf("[GetParcelsList] failed to fetch parcel list Error: %v", err)
+			return []model.Parcel{}, fmt.Errorf("parcel list for offset %d is not found. :%w", offset, sql.ErrNoRows)
+		}
+		return []model.Parcel{}, err
```
Comment (shuza) [inline internal/app/parcel/repository.go]: you can return `nil` instead of empty array

File: internal/app/parcel/repository_test.go [hunk 37]
```diff
@@ -67,12 +68,124 @@ func TestRepository_InsertParcel(t *testing.T) {
...
+		offset = -1
+
+		m.ExpectQuery("^SELECT (.+) FROM parcel WHERE (.+)").
+			WithArgs(0, 0, -1).
+			WillReturnError(sql.ErrNoRows)
+
+		repo := NewRepository(sqlxDB)
+		_, err := repo.GetParcelsList(context.Background(), status, limit, offset)
+
+		assert.EqualError(t, err, "parcel list for offset -1 is not found. :"+sql.ErrNoRows.Error())
```
Comment (shuza) [inline internal/app/parcel/repository_test.go]: use `fmt.Sprintf` to build string instead of `+`

**Review by shuza — CHANGES_REQUESTED** [review]
use string builder instead of `+` if possible

File: internal/app/parcel/repository.go [hunk 38]
```diff
@@ -49,6 +50,18 @@ func (r *repository) InsertParcel(ctx context.Context, parcel model.Parcel) erro
 	return nil
 }
 
+func (r *repository) GetParcelsList(ctx context.Context, status int, limit int, offset int) ([]model.Parcel, error) {
+	var parcels []model.Parcel
+	if err := r.db.SelectContext(ctx, &parcels, getParcelListQuery, status, limit, offset); err != nil {
+		if err == sql.ErrNoRows {
+			log.Error().Err(err).Msgf("[GetParcelsList] failed to fetch parcel list Error: %v", err)
+			return []model.Parcel{}, fmt.Errorf("parcel list for offset %d is not found. :%w", offset, sql.ErrNoRows)
```
Comment (shuza) [inline internal/app/parcel/repository.go]: error should be `nil`

**Review by rdhawladar — APPROVED** [review]

**Review by zrshishir — APPROVED** [review]

**Comment by zrshishir** [comment]
looks good to me

**Review by shuza — APPROVED** [review]
